doc/design/service-discovery.md

```diff
@@ -0,0 +1,261 @@
+# Service Discovery
+
+## Summary and problem description
+
+Service discovery over IPC channel e.g. message queue or UNIX domain socket is not performant since larger data is
+transferred, which can lead to transmission of several frames. If lots of services are discovered at high-frequency
+e.g. at startup the IPC channel can become a bottleneck. Furthermore, a lesson learned from the iceoryx development
+so far is, that the IPC channel should just be used for the earliest communication during startup and not for the
+creation of objects inside the shared memory (see [this issue](https://github.com/eclipse-iceoryx/iceoryx/issues/611)).
+
+### Status quo in iceoryx Almond
+
+#### Runtime
+
+* `cxx::expected<InstanceContainer, FindServiceError> PoshRuntime::findService(capro::ServiceDescription)`
+  * Sends message over IPC channel
+
+#### RouDi
+
+* `runtime::IpcMessage {ProcessManager,PortManager}::findService(capro::ServiceDescription)`
+  * Called after RouDi has received a request from the runtime
+  * Sends back answer to `PoshRuntime` over IPC channel
+
+## Terminology
+
+## Requirements
+
+* Polling and event notification approach shall both be possible for service discovery
+
+### Considerations and use-cases
+
+#### AUTOSAR Adaptive
+
+* findService() call as implemented in Almond v1.0
+* ara::com service discovery callbacks
+
+1. `StartFindService(std::function<void(ServiceContainer<T>,ServiceHandle)>, InstanceId)`: Register callback, std::function interface
+1. `StopFindService(std::function<void(ServiceContainer<T>,ServiceHandle)>)`: Unregister callback
+1. `FindService(InstanceId)`: one-shot, polling, synchronous
+
+#### SOME/IP-SD
+
+* Similar to AUTOSAR Adaptive
+
+#### Cyclone DDS and ROS 2
+
+* Cyclone DDS calls this feature [topic discovery](https://github.com/eclipse-cyclonedds/cyclonedds/blob/master/docs/dev/topic_discovery.md). See also the [feature request](https://github.com/eclipse-cyclonedds/cyclonedds/issues/42).
+
+* Based on SDEP (simple endpoint discovery protocol)
+* `dds_ktopic`
+  * stores pointer to `ddsi_topic_definition`
+* Has built-in DCPSTopic topic and contains
```

Comment:
@dennis-adlink @eboasson

IIUC the `DCPSTopic` contains only the delta and not the full registry with all currently available topics?

Reply:
Yes, that's correct. For each locally created topic or discovered remote topic (if enabled), a `DCPSTopic` sample is created (or, more precisely, a sample is lazily created based on the local topic administration when reading `DCPSTopic` data). The `DCPSTopic` sample contains meta-data for that single topic (see DDS spec 2.2.5 for the full list of fields). On deletion of a topic (possibly remote via SEDP), the instance for that topic is disposed.